[backends] Add functionality to TRT backend

gs-olive opened the pull request:
- Add argument parsing for backend arguments to be passed to Torch-TRT
- Add capability to specify IR and other Torch-TRT fields via command line interface
- Add functionality to compilation path and clean up code

cc: @narendasan

xuzhao9 reviewed (approved):
LGTM. Just curious will you be interested in setting up a CI to maintain this?

gs-olive commented:
Thank you for the review. I think we would be interested in setting up a CI to maintain it. @narendasan - what are your thoughts on this?

gs-olive commented:
@xuzhao9 - what would you need from us to set up a CI to maintain the TRT backend?

xuzhao9 commented:
@gs-olive Here is the guide on how to develop a CI: https://github.com/pytorch/benchmark/blob/main/userbenchmark/ADDING_USERBENCHMARKS.md

gs-olive commented:
Hi @xuzhao9 - thanks for the reference. I've added functionality to this PR in a new commit which adds the `userbenchmark` folder and adapts functionality from the main [`run.py`](https://github.com/pytorch/benchmark/blob/745644f391b4d11da107b2c82fe2d7a3eacf561d/run.py). I also added a `ci.yaml` file. I wasn't sure how the benchmarks are invoked in CI, however. My intent was to have this sort of usage:

```bash
# Invokes a singular model
python run_benchmark.py torch_trt --model resnet18 --precision fp32 --bs 4

# Invokes all models in the directory
python run_benchmark.py torch_trt --precision fp32 --bs 4
```

xuzhao9 commented on torchbenchmark/util/backends/trt.py:
```diff
@@ -31,27 +73,63 @@ def _fx2trt():
+
+        trt_input = [
+            torch_tensorrt.Input(shape=input_.shape, dtype=input_.dtype)
+            for input_ in example_inputs
```
Is it necessary to add check that the type of `example_inputs` is `List[tensor]`? Actually, many models have different types of inputs.

xuzhao9 replied:
One suggestion is to use pytree to traverse the input and cast them to `torch_tensorrt.Input`. Similar to this: https://github.com/pytorch/benchmark/blob/9d84f9e6edcae65b43604e5781c828a05121df78/torchbenchmark/util/env_check.py#L186

gs-olive replied:
Since the different `ir` choices for Torch-TRT process inputs differently, but all can handle Torch Tensor inputs, I passed the example inputs directly to the compiler instead, so the selected `ir` can handle the inputs/casting as necessary.

gs-olive commented:
@xuzhao9 - Thank you for the detailed comments and review. This sounds like a good idea to me; I have removed the `ci.yaml` file as requested, and I will follow up with a subsequent PR addressing the remaining comments.

xuzhao9 reviewed (approved):
LGTM, please address my inline comment about checking the input type of `example_inputs`, thanks!

gs-olive commented:
I have addressed the comment regarding example inputs here: https://github.com/pytorch/benchmark/pull/1753#discussion_r1271062716

xuzhao9 reviewed (approved):
LGTM, thanks!
